### PR TITLE
support realm refresh()

### PR DESCRIPTION
**Pull request by blagoev**
Add support for Realm.refresh() and onRefresh callbacks
#793

### CODE REVIEW

File: lib/src/realm_class.dart
Comment (nirinchev) [inline lib/src/realm_class.dart:556]: Why is this taking an action? I'd imagine an API like `Future<void> refreshAsync()` and if you want to do something after the refresh you use:

```
await realm.refreshAsync();
action();
```

It seems redundant to be exposing this as both a promise and a method that takes a callback.

Comment (blagoev) [reply lib/src/realm_class.dart]: This is not refreshAsync implementation but rather a add_refresh_callback implementation. This takes a callback which will get executed only once and returns a Future since you might want to await it or most probably not. Just register the callback.

File: CHANGELOG.md
Comment (desistefanova) [inline CHANGELOG.md:12]: Why not to mention also `refreshAsync`?

Comment (desistefanova) [inline CHANGELOG.md:12]: Why not to mention also `refreshAsync`?

File: test/realm_test.dart
Comment (desistefanova) [inline test/realm_test.dart:1594]: This test looks strange to me. Why "write onRefresh failed" when called == true? Is it expected the Person not to be stored in the realm?

File: test/realm_test.dart
Comment (desistefanova) [inline test/realm_test.dart:1584]: Is it correct to relly on the delay for 1 milisecond to ensure that the `refreshAsync` is completed?

File: test/realm_test.dart
Comment (desistefanova) [inline test/realm_test.dart:1634]: Probably you could use `waitFutureWithTimeout`.

**Review by desistefanova — COMMENTED** [review]
I have added some questions.

File: test/realm_test.dart
Comment (blagoev) [reply test/realm_test.dart:1594]: This is a message if the expect fails. Not that it expect it to fail

File: CHANGELOG.md
Comment (blagoev) [reply CHANGELOG.md:12]: will add it

File: test/realm_test.dart
Comment (blagoev) [reply test/realm_test.dart:1584]: yes, this makes sure to release the event loop.

File: test/realm_test.dart
Comment (blagoev) [reply test/realm_test.dart:1604]: yes since it releases the event loop.

File: test/realm_test.dart
Comment (blagoev) [reply test/realm_test.dart:1634]: This is test specific. The problem with these functions is they become part of the internal testing API. And here problem is highlighted. I prefer to use the common Future API instead of increasing the size of the internal API which then one needs to remember that it exists

File: lib/src/native/realm_core.dart
Comment (nielsenko) [inline lib/src/native/realm_core.dart:771]: Should we await this fix before merging?

File: lib/src/native/realm_core.dart
Comment (nielsenko) [inline lib/src/native/realm_core.dart:775]: Maybe not for this PR, as this is a pattern we have used all over `realm_core.dart`..

Anyway.. Why not make this 
```suggestion
  Future<void> realmRefreshAsync(Realm realm) async {
```
and await the `completer.future` at the end. That way we are sure that any exceptions raised by this functions before that point is correctly wrapped in a `Future.error(ex)`

**Review by desistefanova — APPROVED** [review]

File: test/realm_test.dart
Comment (nielsenko) [inline test/realm_test.dart:1636]:
```suggestion
    var realm = getRealm(Configuration.local([Person.schema]));
    realm = realm.freeze();

    expect(realm.refreshAsync().timeout(const Duration(milliseconds: 10)), completes);
```

File: test/realm_test.dart
Comment (nielsenko) [inline test/realm_test.dart:1626]:
```suggestion
  final realm = getRealm(Configuration.local([Person.schema]));

    final first = expectLater(realm.refreshAsync(), completes);
    realm.beginWrite().commit();
    await (first);

    final second = expectLater(realm.refreshAsync(), completes);
    realm.write(() {});
    await (second);
```

File: test/realm_test.dart
Comment (nielsenko) [inline test/realm_test.dart:1604]:
```suggestion
    final realm = getRealm(Configuration.local([Person.schema]));
    final transaction = await realm.beginWriteAsync();
    expectLater(realm.refreshAsync(), completes);
    realm.add(Person("name"));
    await transaction.commitAsync();
```

File: test/realm_test.dart
Comment (nielsenko) [inline test/realm_test.dart:1594]:
```suggestion
    final realm = getRealm(Configuration.local([Person.schema]));
    final transaction = realm.beginWrite();
    final first = expectLater(realm.refreshAsync(), completes, reason: "beginWrite onRefresh failed");
    realm.add(Person("name"));
    transaction.commit();
    await first;

    realm.write(() {
      expectLater(realm.refreshAsync(), completes, reason: "write onRefresh failed");
      realm.add(Person("name"));
    });
```

**Review by nielsenko — CHANGES_REQUESTED** [review]
Mostly test related suggestions

File: lib/src/native/realm_core.dart
Comment (blagoev) [reply lib/src/native/realm_core.dart:771]: It seems we will anyway, but merging this without it also will work

File: lib/src/native/realm_core.dart
Comment (blagoev) [reply lib/src/native/realm_core.dart:775]: I was debating to the same but for the public API which is the entry point. so `Realm.refreshAsync`. Let's discuss that and we can do it for all. But not in this PR :)

File: test/realm_test.dart
Comment (blagoev) [reply test/realm_test.dart:1636]: I will use the suggestion

File: test/realm_test.dart
Comment (blagoev) [reply test/realm_test.dart:1626]: This changes the test too much as it is not readable what is happening and what is the expectation. Lets leave it as I have written so it is more clear what goes after the other.

File: test/realm_test.dart
Comment (blagoev) [reply test/realm_test.dart:1604]: As above

File: test/realm_test.dart
Comment (blagoev) [reply test/realm_test.dart:1594]: as above

File: test/realm_test.dart
Comment (nielsenko) [reply test/realm_test.dart:1626]: I believe it is clearer like I write it. It is also shorter. I would like us to stop using the flag based approach, and spread the knowledge of how to do it right.

File: test/realm_test.dart
Comment (nielsenko) [inline test/realm_test.dart:1636]: I think you introduced an error here. It should be:
```suggestion
    realm = realm.freeze();

    expect(realm.refreshAsync().timeout(const Duration(milliseconds: 10)), completes);
```
you are not setting the flag anymore, and you are not expecting `completes`

File: test/realm_test.dart
Comment (nirinchev) [reply test/realm_test.dart:1626]: I'm not sure what this test is supposed to test (in either of the forms). `realm.refreshAsync()` should complete synchronously when we're already at the latest version, which is pretty much always when using a single isolate. So it will have resolved already before we even get to `realm.write`. If we want to verify that it actually does something, we need to spin up a second isolate, open the same realm, write a bunch of data in it, then commit the transaction, signal the first isolate, then assert the data is not visible, call refresh async, assert that the data is visible.

Comment (blagoev) [reply test/realm_test.dart:1626]: this test is the same as what we have in Core. Just using Dart

**Comment by blagoev** [comment]
blocked on https://github.com/realm/realm-core/issues/6075

File: test/realm_test.dart
Comment (nielsenko) [reply test/realm_test.dart:1626]: @nirinchev Currently calling `await realm.refreshAsync()` today will block until something is actually written to the realm. So it tests that.. which may well be a bug in the C-api.

Comment (nirinchev) [reply test/realm_test.dart:1626]: Yeah, it's not the intended behavior though. There's no obvious situation in which this is going to be useful behavior.

Comment (nielsenko) [reply test/realm_test.dart:1626]: I can see value in knowing when the next transaction completes, but I agree that is not what this API is intended for.

Comment (desistefanova) [reply test/realm_test.dart]: The test suggested by @nirinchev is added.

File: test/realm_test.dart
Comment (desistefanova) [reply test/realm_test.dart]: Outdated.

File: test/realm_test.dart
Comment (desistefanova) [reply test/realm_test.dart]: Outdated

File: test/realm_test.dart
Comment (desistefanova) [reply test/realm_test.dart]: Outdated.

File: lib/src/realm_class.dart
Comment (blagoev) [inline lib/src/realm_class.dart:568]:
```suggestion
  /// If another process or [Isolate] has made changes to the realm file, this causes
```

File: lib/src/realm_class.dart
Comment (blagoev) [inline lib/src/realm_class.dart:571]: This drastically differs from what .NET has for example. Here is the doc API .NET has currently
https://www.mongodb.com/docs/realm-sdks/dotnet/latest/reference/Realms.Realm.html#Realms_Realm_Refresh

> Note that this may return true even if no data has actually changed.

Also this is not what the native code is doing. Native code is returning `did_refresh` and it does not return `true` if any data actually changed

File: lib/src/realm_class.dart
Comment (blagoev) [inline lib/src/realm_class.dart:579]: same as above

**Review by nielsenko — COMMENTED** [review]
I think we should consider dropping the synchronous version, and renaming the async version to just `refresh`

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1764]: what does it mean `into sync transaction`. This test name is not clear. Also it is named `Realm.refresh()` but executes `refreshAsync`
Should we rename to 
```suggestion
  test('Realm.refreshAsync() sync transaction', () async {
```

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1777]:
```suggestion
    expect(isRefreshed, false, reason: "There is no new data");
```

Comment (blagoev) [reply test/realm_test.dart:1777]: I think the reason should state why we have this expecations. So this should be 
"There should be no new data"

Comment (blagoev) [reply test/realm_test.dart:1777]: In general adding a reason like this here is not adding much info. I see using a `reason` when the expectation is not clear enough

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1778]:
```suggestion
    expect(called, true);
```

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1782]:
```suggestion
  test('Realm.refreshAsync from within a write block', () async {
```

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1795]:
```suggestion
    expect(isRefreshed, false);
```

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1796]:
```suggestion
    expect(called, true);
```

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1800]:
```suggestion
  test('Realm.refreshAsync from within an async transaction', () async {
```

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1811]: remove `reason`

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1812]: remove `reason`

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1816]:
```suggestion
  test('Realm.refreshAsync outside a transaction', () async {
```

Comment (blagoev) [inline test/realm_test.dart:1816]: This test is not need imo. Or the name is really not explaining what we are testing in the test.

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1833]:
```suggestion
  test('Realm.refreshAsync on frozen realm should be no-op', () async {
```

Comment (blagoev) [reply test/realm_test.dart:1833]: why are we testing with refreshAsync in all tests? Why not use the sync version here?

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1841]:
```suggestion
  test('Realm.refresh', () async {
```

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1842]: I would move the completer creation close to where it is first used

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1851]: This test seems unnecessary complex. Why are we sending booleans and waiting for 100 mils?

Comment (blagoev) [reply test/realm_test.dart:1851]: I would expect this to work as well

```dart
 final externalRealm = Realm(Configuration.local([Person.schema], path: path));
      externalRealm.write(() {
        externalRealm.add(Person(personName));
      });
    externalRealm.close();
```

**Review by blagoev — COMMENTED** [review]
[CHANGES REQUESTED]. 
I can't not make my own PRs RED in GH

File: test/realm_test.dart
Comment (desistefanova) [reply test/realm_test.dart]: I removed `Isolate.run` because it doesn't pass for MacOS where we are testing with Flutter 3.3.10.

File: lib/src/native/realm_core.dart
Comment (desistefanova) [reply lib/src/native/realm_core.dart:584]: It is by purpose. We may need to enable the auto refresh in some tests after it has been disabled.

File: test/realm_test.dart
Comment (desistefanova) [reply test/realm_test.dart:1851]: You suggested to close the second realm after the refresh was called. That's why I close it a little bit later. If you think it is not need I can simplify the test.

File: test/realm_test.dart
Comment (desistefanova) [reply test/realm_test.dart]: I agree. Shall I remove it?

File: lib/src/native/realm_core.dart
Comment (blagoev) [reply lib/src/native/realm_core.dart:584]: We don't need to currently. Are you planning on adding more tests?

File: test/realm_test.dart
Comment (blagoev) [reply test/realm_test.dart:1849]: I think it should not be needed if we disable autorefresh. My suggestion was how to workaround autorefresh. Which is not needed now

File: lib/src/native/realm_core.dart
Comment (desistefanova) [reply lib/src/native/realm_core.dart:584]: OK

File: test/realm_test.dart
Comment (desistefanova) [reply test/realm_test.dart:1832]: Done.

File: lib/src/realm_class.dart
Comment (blagoev) [inline lib/src/realm_class.dart:579]:
```suggestion
  /// Note that this may return `true` even if no data has actually changed.
```

File: test/realm_test.dart
Comment (blagoev) [inline test/realm_test.dart:1833]: Can we use `Isolate.run` which returns a future with a result we can `await` here. Then we can remove the recieving port and the weird await for below?

**Review by blagoev — COMMENTED** [review]
[SUGGESTED CHANGES]

**Review by nielsenko — APPROVED** [review]

**Pull request merged**
